File: firstapp/samples/fog/durability.rb
#!/usr/bin/env ruby
require 'fog/openstack'
require 'digest/md5'
require 'net/http'
require 'json'

# step-1
auth_username = "your_auth_username"
auth_password = "your_auth_password"
auth_url      = "http://controller:5000"
project_name  = "your_project_name_or_id"

swift = Fog::Storage::OpenStack.new openstack_auth_url:     auth_url + "/v3/auth/tokens",
                                    openstack_domain_id:    "default",
                                    openstack_username:     auth_username,
                                    openstack_api_key:      auth_password,
                                    openstack_project_name: project_name

# step-2
container_name = "fractals"
container      = swift.directories.create key: container_name

p container

# step-3
p swift.directories.all

# step-4
file_path   = "goat.jpg"
object_name = "an amazing goat"
container   = swift.directories.get container_name
object      = container.files.create body: File.read(File.expand_path(file_path)),
                                     key:  object_name

# step-5
p container.files.all

# step-6
p container.files.get object_name

# step-7
puts Digest::MD5.hexdigest(File.read(File.expand_path(file_path)))

# step-8
object.destroy

# step-9
p container.files.all

# step-10
container_name = 'fractals'
container      = swift.directories.get container_name

# step-11
endpoint  = "http://IP_API_1"
uri       = URI("#{endpoint}/v1/fractal")
uri.query = URI.encode_www_form results_per_page: -1
data      = JSON.parse(Net::HTTP.get_response(uri).body)

data["objects"].each do |fractal|
  body   = open("#{endpoint}/fractal/#{fractal["uuid"]}") {|f| f.read}
  object = container.files.create body: body, key: fractal["uuid"]
end

p container.files.all

# step-12
container.files.each do |file|
  file.destroy
end
container.destroy

# step-13
object_name = "backup_goat.jpg"
file_path   = "backup_goat.jpg"
extra = {
  description: "a funny goat",
  created:     "2015-06-02"
}
object  = container.files.create body: File.read(File.expand_path(file_path)),
                                 key:  object_name,
                                 metadata: extra

# step-14
#TBC

# step-15

File: firstapp/samples/libcloud/durability.py
# step-1
from __future__ import print_function
from libcloud.storage.types import Provider
from libcloud.storage.providers import get_driver

auth_username = 'your_auth_username'
auth_password = 'your_auth_password'
auth_url = 'http://controller:5000'
project_name = 'your_project_name_or_id'
region_name = 'your_region_name'

provider = get_driver(Provider.OPENSTACK_SWIFT)
swift = provider(auth_username,
                 auth_password,
                 ex_force_auth_url=auth_url,
                 ex_force_auth_version='2.0_password',
                 ex_tenant_name=project_name,
                 ex_force_service_region=region_name)

# step-2
container_name = 'fractals'
container = swift.create_container(container_name=container_name)
print(container)

# step-3
print(swift.list_containers())

# step-4
file_path = 'goat.jpg'
object_name = 'an amazing goat'
container = swift.get_container(container_name=container_name)
object = container.upload_object(file_path=file_path, object_name=object_name)

# step-5
objects = container.list_objects()
print(objects)

# step-6
object = swift.get_object(container_name, object_name)
print(object)

# step-7
import hashlib
print(hashlib.md5(open('goat.jpg', 'rb').read()).hexdigest())

# step-8
swift.delete_object(object)

# step-9
objects = container.list_objects()
print(objects)

# step-10
container_name = 'fractals'
container = swift.get_container(container_name)

# step-11
import json

import requests

endpoint = 'http://IP_API_1'
params = { 'results_per_page': '-1' }
response = requests.get('%s/v1/fractal' % endpoint, params=params)
data = json.loads(response.text)
for fractal in data['objects']:
    response = requests.get('%s/fractal/%s' % (endpoint, fractal['uuid']), stream=True)
    container.upload_object_via_stream(response.iter_content(), object_name=fractal['uuid'])

for object in container.list_objects():
    print(object)

# step-12
for object in container.list_objects():
    container.delete_object(object)
swift.delete_container(container)

# step-13
file_path = 'goat.jpg'
object_name = 'backup_goat.jpg'
extra = {'meta_data': {'description': 'a funny goat', 'created': '2015-06-02'}}
with open('goat.jpg', 'rb') as iterator:
    object = swift.upload_object_via_stream(iterator=iterator,
                                            container=container,
                                            object_name=object_name,
                                            extra=extra)

# step-14
swift.ex_multipart_upload_object(file_path, container, object_name,
                                 chunk_size=33554432)

# step-15
